cpp/src/strings/json/json_to_array.cu

```diff
@@ -0,0 +1,563 @@
+{
+  for (size_t idx = 0; idx < num_chars; idx++) {
+    if (str1[idx] != str2[idx]) { return false; }
+  }
+  return true;
+}
+
+
+CUDA_HOST_DEVICE_CALLABLE bool is_digit(char c)
+{
+  if (c >= '0' && c <= '9') return true;
+  return false;
+}
+
+struct json_string {
+  const char* str;
+  int64_t len;
+
+  CUDA_HOST_DEVICE_CALLABLE bool operator==(json_string const& cmp)
+  {
+    return len == cmp.len && str != nullptr && cmp.str != nullptr &&
+           device_strncmp(str, cmp.str, static_cast<size_t>(len));
+  }
+};
+
+struct pre_compute_size {
+  size_t num_rows; 
+  size_t key_output_len;
+  size_t value_output_len;
+  size_t num_lists;
+};
+
+enum first_operator_type { NONE, OBJECT, ARRAY };
+struct json_first_operator {
+  first_operator_type type;
+  json_string name;
+  int index;
+};
+
+
+
+struct json_array_output {
+  size_t output_max_len;
+  size_t output_len;
+  size_t num_rows;
+  char* output; 
+  size_type* offset;
+  size_type* list_offsets;
+
+  CUDA_HOST_DEVICE_CALLABLE void add_output(const char* str, size_t len)
+  {
+    if (output != nullptr) {
+      // assert output_len + len < output_max_len
+      memcpy(output + output_len, str, len);
+    }
+    output_len += len;
+    num_rows++;
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE void add_output(json_string str) { add_output(str.str, str.len); }
+};
+
+class json_state : private parser {
+ public:
+  CUDA_HOST_DEVICE_CALLABLE json_state()
+    : parser(), element(first_operator_type::NONE), cur_el_start(nullptr)
+  {
+  }
+  CUDA_HOST_DEVICE_CALLABLE json_state(const char* _input, int64_t _input_len)
+    : parser(_input, _input_len), element(first_operator_type::NONE), cur_el_start(nullptr)
+  {
+  }
+  CUDA_HOST_DEVICE_CALLABLE json_first_operator get_first_character()
+  {
+    char c = parse_char();
+    switch (c) {
+      case '{': {
+        json_first_operator op;
+        json_string j_obj{"{", 1};
+        op.name = j_obj;
+        op.type = OBJECT;
+        element = OBJECT;
+        return op;
+      } break;
+      case '[': {
+        json_first_operator op;
+        json_string j_arr{"[", 1};
+        op.name = j_arr;
+        op.type = ARRAY;
+        element = ARRAY;
+        return op;
+      }
+      default: break;
+    }
+    return {NONE};
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE json_string extract_key()
+  {
+    json_string key;
+    if (element == ARRAY) {
+      parse_whitespace();
+      if (*pos == '{')
+        pos++;
+    }
+    parse_string(key, true);
+    return key;
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE char const* find_char_pos(const char ch)
+  {
+    char const* idx = pos;
+    while (*idx != ch && !eof(idx))  idx++;
+    return idx;
+  }
+  CUDA_HOST_DEVICE_CALLABLE char const* find_last_validity_pos()
+  {
+    char const* last_validity = pos;
+    while (is_digit(*last_validity) || *last_validity == '.')  last_validity++;
+    return last_validity;
+  }
+  CUDA_HOST_DEVICE_CALLABLE json_string extract_value(bool& finish_obj, bool& finish_arr)
```

Comment:
`CUDA_HOST_DEVICE_CALLABLE` is used everywhere. I'm not sure if the functions here are supposed to run on the GPU, as the functions involve a lot of `if else`, which is extremely detrimental for CUDA performance.